code/dictation.py

```diff
@@ -1,102 +1,229 @@
+        # whitespace is not affected by formatter state, if text.isspace() is
+        # True we don't need context-sensitivity.
+        do_the_dance = (setting_context_sensitive_dictation.get()
+                        and not text.isspace())
+        if do_the_dance:
+            dictation_formatter.update_context(
+                actions.user.dictation_peek_left(clobber=True))
+        text = dictation_formatter.format(text)
+        actions.user.add_phrase_to_history(text)
+        actions.insert(text)
+        # Add a space after cursor if necessary.
+        if not do_the_dance or not text or text[-1] in no_space_after:
+            return
+        char = actions.user.dictation_peek_right()
+        if char is not None and needs_space_between(text, char):
+            actions.insert(" ")
+            actions.edit.left()
+
+    def dictation_peek_left(clobber: bool = False) -> Optional[str]:
+        """
+        Tries to get some text before the cursor, ideally a word or two, for the
+        purpose of auto-spacing & -capitalization. Results are not guaranteed;
+        dictation_peek_left() may return None to indicate no information. (Note
+        that returning the empty string "" indicates there is nothing before
+        cursor, ie. we are at the beginning of the document.)
+
+        If there is currently a selection, dictation_peek_left() must leave it
+        unchanged unless `clobber` is true, in which case it may clobber it.
+        """
+        # Get rid of the selection if it exists.
+        if clobber: actions.user.clobber_selection_if_exists()
+        # Otherwise, if there's a selection, fail.
+        elif "" != actions.edit.selected_text(): return None
+
+        # In principle the previous word should suffice, but some applications
+        # have a funny concept of what the previous word is (for example, they
+        # may only take the "`" at the end of "`foo`"). To be double sure we
+        # take two words left. I also tried taking a line up + a word left, but
+        # edit.extend_up() = key(shift-up) doesn't work consistently in the
+        # Slack webapp (sometimes escapes the text box).
+        actions.edit.extend_word_left()
+        actions.edit.extend_word_left()
+        text = actions.edit.selected_text()
+        # if we're at the beginning of the document/text box, we may not have
+        # selected any text, in which case we shouldn't move the cursor.
+        if text:
+            # Unfortunately, in web Slack, if our selection ends at newline,
+            # this will go right over the newline. Argh.
+            actions.edit.right()
+        return text
 
-    def auto_format_resume():
-        """Resumes the autoformatter"""
-        return auto_formatter.pause(False)
+    def clobber_selection_if_exists():
+        """Deletes the currently selected text if it exists; otherwise does nothing."""
+        actions.key("space backspace")
```

Comment:
A lot of the flakiness I'm observing in various applications is not actually the clipboard, but the fact that talon presses keys so fast. 

peek left is basically equivalent to
`actions.key("space backspace shift-alt-left shift-alt-left")` on Mac. This reliably fails in applications like Messages.

Potential fixes:
1.  check whether there's a selection before clobber, which is slow and may fail under load
2. Add a delay here. Fragile.
3. Go nuclear: we could consider making this context sensitive dictation only available in the draft window.

Reply:
4. Tinker with the talon's key_wait stuff per-app